Step4 구간 제거 기능

sueH237 opened the pull request:
태경리뷰어님~ 안녕하세요~😄😄 
Step4 구간 제거 기능 PR 요청드립니다.
말씀해주신 Step3의 피드백 내용도 반영하였습니다.

ATDD 미션의 마지막 Step이네요. 
매번 친절한 리뷰와 말씀해주신 피드백으로 성장할 수 있었습니다.

이번에도 시간 되실 때 리뷰 부탁드립니다.
감사합니다~ 😄

sueH237 committed:
docs: 구간 제거 기능 목록 작성

sueH237 committed:
feat: step3 pr 리뷰 반영

- line 접근제어자 수정
- 외래키 설정 추가
- optional=false 선언
- 주석 제거
- Sections 도메인 테스트 추가

sueH237 committed:
feat: 구간 삭제 station 이 middle 인 경우, 인수 테스트와 구현

sueH237 committed:
feat: 구간 삭제 예외인 경우 구현과 인수 테스트 구현

sueH237 committed:
refactor: 원시값 포장 및 메소드 분리

sueH237 committed:
test: 도메인 test 추가

ohtaeg commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -73,4 +73,11 @@ private Section toSection(SectionRequest sectionRequest) {
+
+    @Transactional
+    public void deleteSectionByStationId(Long lineId, Long stationId) {
+        Line line = lineRepository.findById(lineId).orElseThrow(() -> new NoSuchElementException(ErrorMessage.ERROR_LINE_NOT_EXIST));
+        Station station = findStationById(stationId);
```
꼼꼼하게 잘 체크해주셨네요 👍

ohtaeg commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -12,19 +12,19 @@ public class Section implements Comparable<Section>{
 
-    @ManyToOne(fetch = FetchType.LAZY)
-    @JoinColumn(name = "upStation_id")
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "upStation_id", foreignKey = @ForeignKey(name = "fk_up_station_section_to_station"))
```
피드백 반영 👍

ohtaeg commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -94,4 +94,17 @@ private void checkDistance(Section newSection) {
+        return station.equals(this.downStation);
+    }
+
+    public void refreshWith(Section upSection) {
```
개인적으로 외부로 공개된 refreshWith 이라는 뜻은 무엇을 refresh 한다는 것인지 직관적으로 알기 힘든 것 같아요 😭 
좀 더 구체적인 명명도 괜찮지 않을까 합니다 😄

sueH237 replied:
구체적인 명명(changeDownStationDistance)으로 변경하였습니다~ 😄😄 직관적으로 알기 쉬운 메소드 명을 작성해야 하는데, 알려주셔서 감사합니다😄😄

ohtaeg commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -39,4 +43,56 @@ public List<Station> getStations() {
+        checkDeleteEnd(upSection, downSection);
+    }
+
+    private void checkSectionInLine(Optional<Section> upSection, Optional<Section> downSection) {
```
이 부분은 같이 고민해보시면 좋을 것 같아요! 
javadoc에서도 Optional은 주로 반환 타입으로 사용되길 위한다고 써있는데요!
```
Optional is primarily intended for use as a method return type where there is a clear need to represent "no result," and where using null is likely to cause errors.
```
뿐더러 소나큐브에서도 Optional의 탄생 배경은 Optional은 반환 타입으로만 사용되도록 의도한 설계라고 하며,
파라미터로 Optional 사용을 지양하라고 하는데요 :)

https://rules.sonarsource.com/java/RSPEC-3553

해당 부분들을 인지하고 Optional 파라미터를 의도하고 사용하신걸까요 :) ?

개인적으로도 메서드의 파라미터가 인스턴스일 때 Optional이든 아니든
null 체크를 메서드 내부에서 항상 하는 것이 언제나 안전하기에
인스턴스를 Optional로 한번 더 감쌀 이유가 있을지 개인적으로 항상 고민을 하는 부분이에요 :)

sueH237 replied:
흠.....고민이 되는 부분입니다..
null check 를 내부에서 할 수 있도록 optional을 파라미터로 사용했던 메서드를 수정하였습니다.
Optional 에 관해서 소나큐브에서 파라미터로 사용은 지양하는 건 이번 기회에 알게 되었습니다.
알려주셔서 감사합니다.👍👍

ohtaeg commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -39,4 +43,56 @@ public List<Station> getStations() {
+
+    private void deleteMiddle(Section upSection, Section downSection) {
+        downSection.refreshWith(upSection);
+        sections.remove(upSection);
```
상태를 변경할 시 최대한 한곳에서 할 수 있도록 하는 것도 괜찮을 것 같아요 😄 
```suggestion
        removeEndSection(upSection);
```

sueH237 replied:
👍👍 한곳에서 할 수 있도록 변경하는 편이 가독성면에서도 좋을 것 같습니다! 자세히 알려주셔서 감사합니다

ohtaeg commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -39,4 +43,56 @@ public List<Station> getStations() {
+        downSection.ifPresent(this::removeEndSection);
+    }
+
+    private void removeEndSection(Section section) {
+        sections.remove(section);
```
메서드는 종점 구간을 삭제하는 의미를 가졌지만 실제 삭제는 위치를 고려하지 않는 것 같아요 ㅎㅎ
네이밍을 변경해보면 어떨까요 :) ?

sueH237 replied:
네이밍 removeSection 으로 변경하였습니다:D

ohtaeg commented on src/test/java/nextstep/subway/section/SectionAcceptanceTest.java:
```diff
@@ -118,6 +118,70 @@ void addSectionNewDownStation() {
+    @Test
+    void removeMiddleStation() {
+        // given
+        StationResponse 새로운역 = createStationRest("새로운역").as(StationResponse.class);
+        SectionRequest newSection = new SectionRequest(새로운역.getId(),하행종점역.getId(),5);
+        addSectionRest(신분당선.getId(), newSection);
+        removeStationRest(신분당선.getId(), 새로운역.getId());
```
인수 테스트마다 반복되는 것 같아요..!
인수 테스트 취지에 맞게 좀 더 가독성 있게 혹은 메서드로 분리해볼 수 있을까요 :) ?

sueH237 replied:
메소드로 분리해보겠습니다😄😄

ohtaeg reviewed (changes requested):
리뷰가 늦어서 죄송합니다 🙏 4단계도 구현을 잘해주셨어요 👍 👍 
소소한 코멘트를 드렸는데 반영이 되면 좋을 것 같아 Request chansges 하였습니다..! 🙏 
마지막 피드백이 될 것 같네요 😭 조금만 더 화이팅입니다 🔥

sueH237 committed:
feat: Step4 pr 리뷰 반영

- 메소드명 구체적으로 명명 수정
- Optional 파라미터 사용 지양하도록 수정
- 상태 변경 한 곳에서 하도록 수정
- 인수테스트 공통 메소드 분리

ohtaeg commented on src/test/java/nextstep/subway/section/SectionAcceptanceTest.java:
```diff
@@ -64,9 +65,7 @@ public void setUp() {
-        StationResponse 새로운역 = createStationRest("새로운역").as(StationResponse.class);
-        SectionRequest newSection = new SectionRequest(상행종점역.getId(),새로운역.getId(),5);
-        addSectionRest(신분당선.getId(), newSection);
+        노선_구간_등록(상행종점역.getId(),새로운역.getId(),7);
```
피드백 반영 👍

ohtaeg reviewed (approved):
마지막까지 피드백을 잘반영해주셨어요 👍 👍
드렸던 피드백들이 조금이나마 수영님께 도움이 되었으면 좋겠네요 :)
이만 merge 하도록 하겠습니다!! 고생 많으셨어요 🙇
남은 미션들도 끝까지 화이팅입니다 🔥
